File: samples/scala/shorturls/app/controllers/ApiController.scala
package controllers

import play.api._
import play.api.mvc._
import models.{IdGenerator, ShortURLs, ShortURL}
import play.api.libs.json.Json
import models.ShortURLs._
import scala.concurrent.Future
import play.api.Play.current
import play.api.libs.{EventSource, Comet}


object ApiController extends Controller {

  def getUrl(id: String) = Action.async {
    ShortURLs.findById(id).map { maybe =>
      maybe.map( url => Ok( Json.toJson(url)) ).getOrElse(NotFound)
    }
  }

  def getAllUrls() = Action.async {
    ShortURLs.findAll().map(urls => Ok( Json.toJson( urls ) ) )
  }

  implicit val toJson = Comet.CometMessage[ShortURL]( url => Json.stringify( ShortURLs.urlWriter.writes( url ) ) )

  def getAllUrlsAsSSE() = Action.async {
    ShortURLs.findAllAsEnumerator().map { enumerator => Ok.chunked( enumerator.through( EventSource() ) ) }
  }

  def pollAll() = Action {
    Ok.chunked( ShortURLs.pollAll().through( EventSource() ) )
  }

  def createUrl() = Action.async { implicit request =>
    ShortURLs.urlForm.bindFromRequest.fold(
      errors => Future(BadRequest(
        Json.obj(
          "status" -> "error",
          "error" -> true,
          "created" -> false,
          "message" -> "You need to pass a non empty url value")
        )
      ),
      url => {
        ShortURLs.findByURL(url).flatMap { maybeUrl =>
          maybeUrl.map { shortUrl =>
            Future(
              Ok(
                Json.obj(
                  "status" -> "existing",
                  "error" -> false,
                  "created" -> true,
                  "message" -> "already exists",
                  "url" -> shortUrl
                )
              )
            )
          }.getOrElse {
            IdGenerator.nextId().flatMap { id =>
              val shortUrl = ShortURL(s"$id", url)
              ShortURLs.save(shortUrl).map { status =>
                status.isSuccess match {
                  case true => Ok(
                    Json.obj(
                      "status" -> "created",
                      "error" -> false,
                      "created" -> true,
                      "message" -> status.getMessage,
                      "url" -> shortUrl
                    )
                  )
                  case false => BadRequest(
                    Json.obj(
                      "status" -> "error",
                      "error" -> true,
                      "created" -> false,
                      "message" -> status.getMessage
                    )
                  )
                }
              }
            }
          }
        }
      }
    )
  }

  def delete(id: String) = Action.async {
    ShortURLs.remove(id).map { status =>
      status.isSuccess match {
        case true => Ok(
          Json.obj(
            "status" -> "deleted",
            "error" -> false,
            "deleted" -> true,
            "message" -> status.getMessage
          )
        )
        case false => BadRequest(
          Json.obj(
            "status" -> "error",
            "error" -> true,
            "deleted" -> false,
            "message" -> status.getMessage
          )
        )
      }
    }
  }
}

File: samples/java/shorturls/app/controllers/ApiController.java
package controllers;

import models.IdGenerator;
import models.ShortURL;
import net.spy.memcached.ops.OperationStatus;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.reactivecouchbase.client.OpResult;
import play.libs.F;
import play.libs.Json;
import play.mvc.Controller;
import play.mvc.Result;

import java.util.Collection;

public class ApiController extends Controller {

    public static Result getUrl(String id) {
        return async(
            ShortURL.findById(id).map(new F.Function<ShortURL, Result>() {
                @Override
                public Result apply(ShortURL shortURL) throws Throwable {
                    if (shortURL == null) {
                        return notFound();
                    }
                    return ok(Json.toJson(shortURL));
                }
            })
        );
    }

    public static Result delete(String id) {
        return async(
            ShortURL.remove(id).map(new F.Function<OpResult, Result>() {
                @Override
                public Result apply(OpResult status) throws Throwable {
                    if (status.isSuccess()) {
                        ObjectNode node = Json.newObject();
                        node.put("status", "deleted");
                        node.put("error", false);
                        node.put("deleted", true);
                        node.put("message", status.getMessage());
                        return ok(node);
                    } else {
                        ObjectNode node = Json.newObject();
                        node.put("status", "error");
                        node.put("error", true);
                        node.put("deleted", false);
                        node.put("message", status.getMessage());
                        return  badRequest(node);
                    }
                }
            })
        );
    }

    public static Result getAllUrls() {
        return async(
            ShortURL.findAll().map(new F.Function<Collection<ShortURL>, Result>() {
                @Override
                public Result apply(Collection < ShortURL > shortURLs) throws Throwable {
                    return ok(Json.toJson(shortURLs));
                }
            })
        );
    }

    public static Result createUrl() {
        String urlValue = null;
        try {
            urlValue = ShortURL.form.bindFromRequest().get().url;
        } catch(Exception e) {}

        if (urlValue == null) {
            ObjectNode node = Json.newObject();
            node.put("status", "error");
            node.put("error", true);
            node.put("deleted", false);
            node.put("message", "You need to pass a non empty url value");
            return badRequest(node);
        } else {
            final String url = urlValue;
            return async(ShortURL.findByURL(url).flatMap(new F.Function<F.Option<ShortURL>, F.Promise<Result>>() {
                @Override
                public F.Promise<Result> apply(F.Option<ShortURL> maybe) throws Throwable {
                    return maybe.map(new F.Function<ShortURL, F.Promise<Result>>() {
                        @Override
                        public F.Promise<Result> apply(ShortURL shortURL) throws Throwable {
                            ObjectNode node = Json.newObject();
                            node.put("status", "existing");
                            node.put("error", false);
                            node.put("deleted", true);
                            node.put("message", "already exists");
                            node.put("url", Json.toJson(shortURL));
                            return F.Promise.pure((Result) ok(node));
                        }
                    }).getOrElse(new F.Function<Void, F.Promise<Result>>() {
                        @Override
                        public F.Promise<Result> apply(Void aVoid) throws Throwable {
                            return IdGenerator.nextId().flatMap(new F.Function<Object, F.Promise<Result>>() {
                                @Override
                                public F.Promise<Result> apply(Object o) throws Throwable {
                                    Long val = (Long) o;
                                    final ShortURL shortUrl = new ShortURL(val.toString(), url);
                                    return ShortURL.save(shortUrl).flatMap(new F.Function<OpResult, F.Promise<Result>>() {
                                        @Override
                                        public F.Promise<Result> apply(OpResult status) throws Throwable {
                                            if (status.isSuccess()) {
                                                ObjectNode node = Json.newObject();
                                                node.put("status", "created");
                                                node.put("error", false);
                                                node.put("deleted", true);
                                                node.put("message", status.getMessage());
                                                node.put("url", Json.toJson(shortUrl));
                                                return F.Promise.pure((Result) ok(node));
                                            } else {
                                                ObjectNode node = Json.newObject();
                                                node.put("status", "error");
                                                node.put("error", true);
                                                node.put("deleted", false);
                                                node.put("message", status.getMessage());
                                                return F.Promise.pure((Result) badRequest(node));
                                            }
                                        }
                                    });
                                }
                            });
                        }
                    }.apply(null));
                }
            }));
        }
    }
}
